pyanaconda/modules/subscription/installation.py

```diff
@@ -15,32 +15,446 @@
         self._sysroot = sysroot
         self._role = role
         self._sla = sla
         self._usage = usage
         self._addons = addons
+        self.succeeded = Signal()
 
     @property
     def name(self):
         return "Set system purpose"
 
     def run(self):
         system_purpose.give_the_system_purpose(self._sysroot,
-                                                 self._role,
-                                                 self._sla,
-                                                 self._usage,
-                                                 self._addons)
+                                               self._role,
+                                               self._sla,
+                                               self._usage,
+                                               self._addons)
+        self._done()
+
+    @async_action_nowait
+    def _done(self):
+        # tell nayone listening that we are done
+        self.succeeded.emit()
+
+
+class StartRHSMTask(Task):
+    """Start the RHSM DBUS service.
+
+    TODO: this might not be needed once the RHSM DBUS services
+          is a regular part of the installation environment
+    """
+
+    def __init__(self, subscription_url):
+        """Create a new task.
+
+        :param str subscription_url: Red Hat subscription service URL
+        """
+        super().__init__()
+        self._subscription_url = subscription_url
+
+    @property
+    def name(self):
+        return "Start the RHSM service"
+
+    def run(self):
+        """ There is currently a bug in the RHSM service not taking changes from DBUS
+        API into account until service restart, so we need to:
+        - start the RHSM service
+        - set the staging URL
+        - restart the RHSM service
+        After that the RHSM service should be ready for use.
+        """
+        # make sure /etc/yum.repos.d exists
+        # - othwerwise RHSM will not write out the repo file
+        # - this is a know issues and it might be possible to
+        #   drop this code from Anaconda once RHSM handles
+        #   the missing folder better
+        if not os.path.exists("/etc/yum.repos.d"):
+            log.debug("start-rhsm-task: creating /etc/yum.repos.d")
+            os.makedirs("/etc/yum.repos.d")
+
+        # start the service
+        log.debug("start-rhsm-task: starting the RHSM service")
+        util.start_service("rhsm.service")
+
+        time.sleep(5)
+
+        # set the staging URL
+        log.debug("start-rhsm-task: setting main subscription URL to: %s", self._subscription_url)
+        config_proxy = RHSM.get_proxy(RHSM_CONFIG)
+        config_proxy.Set("server.hostname", self._subscription_url, "")
+
+        # restart the service
+        log.debug("start-rhsm-task: restarting RHSM service")
+        util.restart_service("rhsm.service")
+        log.debug("start-rhsm-task: RHSM service should be now ready for use")
+
+
+class RegisterWithUsernamePasswordTask(Task):
```

Comment:
Well, the organization name + activation key are also credentials, so I think it could be ambiguous. Still, the current name is indeed not the best. So maybe `RegisterWithPasswordTask` ? And the same thing for the other one - `RegisterWithActivationKeyTask` ?

Reply:
Seems like a good solution.